packages/mangrove-arbitrage/src/MgvArbitrage2.sol

```diff
@@ -0,0 +1,246 @@
+  ///
+  /// ```
+  /// {
+  ///    outboundTkn: takerWantsToken,
+  ///    inboundTkn: takerGivesToken,
+  ///    gives: x,
+  ///    wants: y,
+  ///    // ... (other properties)
+  /// }
+  /// ```
+  ///
+  /// The primary objective of this function is to exchange a specified amount of `takerGivesToken` (z) to acquire an exact quantity of `takerWantsToken` (y) using uniswap
+  /// and subsequently swap the acquired `takerWantsToken` for `takerGivesToken` using Mangrove.
+  /// After these operations, the function checks that the balances of `takerGivesToken` and `takerWantsToken` have increased. If the balances have not increased, the transaction reverts, indicating that the arbitrage was not profitable.
+  ///
+  /// @param params An `ArbParams` struct containing the necessary information for the arbitrage operation.
+  function doArbitrageFirstUniwapThenMangrove(ArbParams calldata params) public onlyAdmin {
+    uint givesTokenBalance = params.takerGivesToken.balanceOf(address(this));
+
+    uint maxAmount = estimateHowMuchUniswapV3CanGive(params.pool, params.takerGivesToken, givesTokenBalance);
+
+    uint wantsTokenBalance = params.takerWantsToken.balanceOf(address(this));
+
+    MgvStructs.OfferUnpacked memory bestOffer =
+      mangroveGetBestOffer(address(params.takerGivesToken), address(params.takerWantsToken));
+
+    (uint deltaGives, uint deltaWants) = lowLevelUniswapSwap(
+      address(params.takerGivesToken),
+      address(params.takerWantsToken),
+      // compute minimum between my total balance of "takerGivesToken" into "takerWantsToken"
+      // and bestOffer.wants. This permit us to partial fill an offer using first uniswap.
+      bestOffer.wants < maxAmount ? -int(bestOffer.wants) : int(givesTokenBalance),
+      params.pool
+    );
+
+    (uint totalGot, uint totalGave,,) =
+      mgv.marketOrder(address(params.takerGivesToken), address(params.takerWantsToken), 0, deltaWants, false);
+
+    require(params.minimumGain + deltaGives <= totalGot, "MgvArbitrage/notProfitable");
+    require(wantsTokenBalance <= wantsTokenBalance + deltaWants - totalGave, "MgvArbitrage/notProfitable");
+  }
+
+  ///@notice Retrieves the best mangrove offer for a given market pair of outbound token and inbound token onplatformMangrove.
+  ///@param outboundTkn The token sent by the offer maker.
+  ///@param inboundTkn The token sent by the offer taker.
+  ///@return bestOffer A struct containing the best offer parameters.
+  function mangroveGetBestOffer(address outboundTkn, address inboundTkn)
+    internal
+    view
+    returns (MgvStructs.OfferUnpacked memory bestOffer)
+  {
+    uint bestOfferId = mgv.best(outboundTkn, inboundTkn);
+    bestOffer = mgv.offers(outboundTkn, inboundTkn, bestOfferId).to_struct();
```

Comment:
Since you need very little info from `bestOffer` I would not unpack it. Just use `bestOffer.wants()` when needed

Reply:
Fixed